src/com/facebook/buck/shell/DefaultWorkerTool.java

```diff
@@ -85,7 +103,47 @@ public int getMaxWorkers() {
   }
 
   @Override
+  public boolean isPersistent() {
+    return isPersistent;
+  }
+
+  @Override
   public ImmutableSortedSet<BuildRule> getRuntimeDeps() {
     return getDeps();
   }
+
+  @Override
+  public HashCode getInstanceKey() {
+    return buildOutputInitializer.getBuildOutput().getRuleKey().getHashCode();
+  }
+
+  @Override
+  public Data initializeFromDisk(OnDiskBuildInfo onDiskBuildInfo) throws IOException {
+    Optional<RuleKey> ruleKey =
+        onDiskBuildInfo.getRuleKey(BuildInfo.METADATA_KEY_FOR_RULE_KEY);
```

Comment:
Sorry - I misread the question there. We could definitely recompute the rule key to avoid relying on the build info, but I don't know if that's much better. `RuleKeyBuilderFactory` will have to hash the source files anyway, which are likely build outputs and as mutable as the written rule key file. If anything, I think it would help if the data in `OnDiskBuildInfo` wasn't coming directly from disk but rather from the data with which the build was started.